src/DomainNetMap.cpp

```diff
@@ -0,0 +1,265 @@
+    std::unique_ptr<DomainNetMap> DomainNetMap::make_unique(const std::string nn_path, geopm_domain_e domain_type, int domain_index)
+    {
+        return geopm::make_unique<DomainNetMapImp>(nn_path, domain_type, domain_index);
+    }
+
+    DomainNetMapImp::DomainNetMapImp(
+            const std::string nn_path,
+            geopm_domain_e domain_type,
+            int domain_index)
+        : DomainNetMapImp(nn_path, domain_type, domain_index, platform_io(), std::make_shared<NNFactoryImp>()) {
+    }
+
+    DomainNetMapImp::DomainNetMapImp(
+            const std::string nn_path,
+            geopm_domain_e domain_type,
+            int domain_index,
+            PlatformIO &plat_io,
+            std::shared_ptr<NNFactory> nn_factory)
+        : m_platform_io(plat_io)
+          , m_nn_factory(nn_factory)
+    {
+        std::ifstream file(nn_path);
+
+        if (!file) {
+            throw Exception("Unable to open neural net file.\n",
+                    GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        file.seekg(0, std::ios::end);
+        std::streampos length = file.tellg();
+        file.seekg(0, std::ios::beg);
+
+        if (length >= m_max_nnet_size) {
+            throw Exception("Neural net file exceeds maximum size.\n",
+                    GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        std::string buf, err;
+        buf.reserve(length);
+
+        std::getline(file, buf, '\0');
+
+        const json11::Json nnet_json = json11::Json::parse(buf, err);
+
+        // make sure that there no unexpected keys in the json
+        std::vector<std::string> expected_keys = {"layers", "signal_inputs", "delta_inputs", "trace_outputs", "description"};
+        for (auto &key : nnet_json.object_items()) {
+            if (std::find(expected_keys.begin(), expected_keys.end(), key.first) == expected_keys.end()) {
+                throw Exception(
+                        "Unexpected key in neural net json: " + key.first + "\n",
+                        GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+            }
+        }
+
+        if (! nnet_json["layers"].is_array()) {
+            throw Exception(
+                    "Neural net must contain valid json and must have "
+                    "a key \"layers\" whose value is an array.\n",
+                    GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        if (nnet_json["signal_inputs"].array_items().size() == 0 
+            && nnet_json["delta_inputs"].array_items().size() == 0) {
+            throw Exception(
+                    "Neural net json must contain at least one of "
+                    "\"signal_inputs\" and \"delta_inputs\" whose "
+                    "value is a non-empty array.\n",
+                    GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        
+        std::vector<std::shared_ptr<DenseLayer>> layers;
+        for (std::size_t layer_idx = 0; layer_idx < nnet_json["layers"].array_items().size(); ++layer_idx) {
```

Comment:
```suggestion
        for (const auto &layer : nnet_json["layers"]) {
```

Edit line below accordingly.